### PR TITLE
Expose stack tags to programs

### LINKED ISSUE #2144: Expose metadata tags to CLI/programs
Particularly with https://github.com/pulumi/pulumi/issues/109, it would be good to expose the ability to tag stacks and permit querying based on those tags. It turns out we already have this capability and use it for project, repo, and CI metadata, but we simply don't expose it. We should, both in the CLI and possibly even programmatically from inside of a Pulumi program.

### PR DESCRIPTION
**Pull request by justinvp**
Now that we allow managing stack tags via the CLI, expose them to Pulumi programs.

Note that currently only the cloud backend persists tags. This means:

 - When using the local backend, programs will only see the latest
   environment tags.

 - When using the cloud backend, programs will see the stack's tags from
   the service (which includes the latest environment tags that were
   sent to the service when the update started, along with any custom
   tags that have been set).

As a follow-up, I can add support for persisting tags for the local backend. Should be easy enough, just storing the data in a file next to where we store the stack data. It didn't seem as important, but it would be nice to have consistency between the two backends.

Fixes #2144

### CODE REVIEW

File: pkg/resource/deploy/target.go [hunk 1]
```diff
@@ -15,16 +15,18 @@
...
 type Target struct {
-	Name      tokens.QName     // the target stack name.
-	Config    config.Map       // optional configuration key/value pairs.
-	Decrypter config.Decrypter // decrypter for secret configuration values.
-	Snapshot  *Snapshot        // the last snapshot deployed to the target.
+	Name      tokens.QName                    // the target stack name.
+	Config    config.Map                      // optional configuration key/value pairs.
+	Decrypter config.Decrypter                // decrypter for secret configuration values.
+	Snapshot  *Snapshot                       // the last snapshot deployed to the target.
+	Tags      map[apitype.StackTagName]string // the target stack's tags.
```
Comment (justinvp) [inline pkg/resource/deploy/target.go]: This seemed like the best place to flow the stack tags, but let me know if there's a better place to hang these.

File: sdk/nodejs/metadata.ts [hunk 2]
```diff
@@ -36,3 +37,17 @@ export function getStack(): string {
...
+ * getStackTag returns a stack tag's value or undefined if it is unset.
+ */
+export function getStackTag(name: string): string | undefined {
+    return runtime.getStackTag(name);
+}
+
+/**
+ * getStackTags returns a copy of all the stack tags.
+ */
+export function getStackTags(): {[name: string]: string} {
```
Comment (justinvp) [inline sdk/nodejs/metadata.ts]: @lukehoban, We didn't do a formal review of these so just want to make sure you're OK with these two new APIs: `getStackTag` and `getStackTags`. There are similarly named APIs for Go and Python.

Comment (justinvp) [reply sdk/nodejs/metadata.ts]: Node:

```typescript
/**
 * getStackTag returns a stack tag's value or undefined if it is unset.
 */
export function getStackTag(name: string): string | undefined {
    return runtime.getStackTag(name);
}

/**
 * getStackTags returns a copy of all the stack tags.
 */
export function getStackTags(): {[name: string]: string} {
    return runtime.allStackTags();
}
```

Python:

```python
def get_stack_tag(name: str) -> str:
    """
    Returns a stack tag's value or None.
    """
    return runtime.get_stack_tag(name)


def get_stack_tags() -> Dict[str, str]:
    """
    Returns a copy of all the stack tags.
    """
    return runtime.get_stack_tags()
```

Go:

```go
// GetStackTag returns the stack tag value, as a string, and a bool indicating whether it exists or not.
func (ctx *Context) GetStackTag(name string) (string, bool) {
	v, ok := ctx.info.StackTags[name]
	return v, ok
}

// GetStackTags returns a copy of the stack tags.
func (ctx *Context) GetStackTags() map[string]string {
	if ctx.info.StackTags == nil {
		return nil
	}

	// Make a copy of the tags.
	copy := make(map[string]string)
	for k, v := range ctx.info.StackTags {
		copy[k] = v
	}
	return copy
}
```

File: sdk/python/lib/pulumi/metadata.py [hunk 3]
```diff
@@ -27,3 +29,17 @@ def get_stack() -> str:
     Returns the current stack name.
     """
     return runtime.get_stack()
+
+
+def get_stack_tag(name: str) -> str:
+    """
+    Returns a stack tag's value or None.
```
Comment (swgillespie) [inline sdk/python/lib/pulumi/metadata.py]: If this function can return `None`, its type should be `Optional[str]`

Comment (justinvp) [reply sdk/python/lib/pulumi/metadata.py]: Thanks! I'm a Python noob. Fixed.

**Comment by justinvp** [comment]
Closing per https://github.com/pulumi/pulumi/issues/2144#issuecomment-456884635

**Pull request closed without merging**